docs/development.rst

```diff
@@ -26,157 +121,606 @@ Quickstart
+which is imported by the high-level Python code. The low-level Python module is
+not intended to be used directly by users and may change arbitrarily over time.
+
+The usual pattern in the low-level Python API is to define a Python class
+which corresponds to a given "class" in the C API. For example, we define
+a ``TreeSequence`` class, which is essentially a thin wrapper around the
+``tsk_tree_t`` type from the C library.
+
+The ``_tskitmodule.c`` file follows the standard conventions given in the
+`Python documentation <https://docs.python.org/3.6/extending/index.html>`_.
+
+---------
+Compiling
+---------
+
+The ``setup.py`` file describes the requirements for the low-level ``_tskit``
+module and how it is built from source. To build the module so that it is available
+for use in the current working directory, run
+
+.. code-block:: bash
+
+    $ python3 setup.py build_ext --inplace
+
+A development Makefile is also provided in the ``python`` directory, so that running
+``make`` should build the extension module in place.
+
+For C debugging you can force a debug build of the `_tskit` module using
+
+.. code-block:: bash
+
+    CFLAGS='-Wall -O0 -g' make
+
+------------------------
+Testing for memory leaks
+------------------------
+
+The Python C API can be subtle, and it is easy to get the reference counting wrong.
+The ``stress_lowlevel.py`` script makes it easier to track down memory leaks
+when they do occur. The script runs the unit tests in a loop, and outputs
+memory usage statistics.
+
+
+.. _sec_development_continuous_integration:
+
+****************************
+Continuous Integration tests
+****************************
+
+A number of different continuous integration providers are used, which run different
+combinations of tests on different platforms, as well as running various
+checks for code quality.
+
+- A `Github action <https://help.github.com/en/actions>`_ runs some code style and quality checks.
+
+- `CircleCI <https://circleci.com/>`_ runs all Python tests using the apt-get
+  infrastructure for system requirements. We also runs C tests, compiled
+  using gcc and clang, and check for memory leaks using valgrind.
+  Documentation is also built to check for any errors.
+
+- `Travis CI <https://travis-ci.org/>`_ runs Python tests on Linux and OSX using the
+  `Conda <https://conda.io/docs/>`__ infrastructure for the system level
+  requirements.
+
+- `AppVeyor <https://www.appveyor.com/>`_ runs Python tests on
+  Windows using conda.
+
+- `CodeCov <https://codecov.io/gh>`__ tracks test coverage in Python and C.
+
+
+.. _sec_development_best_practices:
 
 ******************************
 Best Practices for Development
 ******************************
 
 The following is a rough guide of best practices for contributing a function to the
 tskit codebase.
-We will use the :meth:`Tree.kc_distance` method as an example. 
+We will use the :meth:`Tree.kc_distance` method as an example.
```

Comment:
This isn't actually used much below - I'd either omit it or put links to the appropriate things in most of the steps.